mindmeld/components/_util.py

```diff
@@ -38,3 +39,207 @@ def _get_module_or_attr(module_name: str, func_name: str = None):
+        # root
+        self.root = TreeNode('root', allow=allow, children=[])
+        # construct NLP tree
+        for domain in nlp.domains:
+            domain_node = TreeNode(domain, parent=self.root, allow=allow, children=[])
+            self.root.children.append(domain_node)
+            for intent in nlp.domains[domain].intents:
+                intent_node = TreeNode(intent, parent=domain_node, allow=allow, children=[])
+                domain_node.children.append(intent_node)
+                entities = nlp.domains[domain].intents[intent].entities
+                for entity in entities:
+                    entity_node = TreeNode(entity, parent=intent_node, allow=allow, children=[])
+                    intent_node.children.append(entity_node)
+                    for role in entities[entity].role_classifier.roles:
+                        role_node = TreeNode(role, parent=intent_node, allow=allow, children=None)
+                        entity_node.children.append(role_node)
+
+    def get_domains(self):
+        return self.root.children or []
+
+    def get_intents(self, domain: Union[str, TreeNode]):
+        if isinstance(domain, TreeNode):
+            domain = domain.value
+
+        for domain_node in self.root.children:
+            if domain_node.value == domain:
+                return domain_node.children
+        return []
+
+    def get_entities(self, domain: Union[str, TreeNode],
+                     intent: Union[str, TreeNode]):
+
+        if isinstance(domain, TreeNode):
+            domain = domain.value
+
+        if isinstance(intent, TreeNode):
+            intent = intent.value
+
+        for intent_node in self.get_intents(domain):
+            if intent_node.value == intent:
+                return intent_node.children
+        return []
+
+    def get_roles(self, domain: Union[str, TreeNode],
+                  intent: Union[str, TreeNode],
+                  entity: Union[str, TreeNode]):
+        if isinstance(domain, TreeNode):
+            domain = domain.value
+
+        if isinstance(intent, TreeNode):
+            intent = intent.value
+
+        if isinstance(entity, TreeNode):
+            entity = entity.value
+
+        for entity_node in self.get_entities(domain, intent):
+            if entity_node.value == entity:
+                return entity_node.children
+        return []
+
+    def update(self, allow: bool, domain: Union[str, TreeNode],
+               intent: Optional[Union[str, TreeNode]] = None,
+               entity: Optional[Union[str, TreeNode]] = None,
+               role: Optional[Union[str, TreeNode]] = None):
+        """
+        This function updates the NLP tree with mask values
+        Args:
+            allow: True is mask off, False is mask on
+            domain: domain of NLP
+            intent: intent of NLP
+            entity: entity of NLP
+            role: role of NLP
+        """
+
+        if isinstance(domain, TreeNode):
```

Comment:
Since this block of code is called a few times, could we instead make a function `convert_tree_node_to_values(List[Union[str,TreeNode]])` which iterates through a list and converts any TreeNode objects to string values?

Reply:
done